packages/runtime/container-runtime/src/summaryManagerUtils.ts

```diff
@@ -0,0 +1,200 @@
+}
+
+interface IOrderedClientsEvents extends IEvent {
+    (event: "currentChange", listener: (client: ITrackedClient | undefined) => void);
+    (event: "summarizerChange", listener: (summarizerCount: number) => void);
+}
+
+export class OrderedClients extends TypedEventEmitter<IOrderedClientsEvents> {
+    private readonly clientMap = new Map<string, IFullClient>();
+    private readonly rootClient: ILinkedClient = {
+        sequenceNumber: -1,
+        clientId: "",
+        prevClient: undefined,
+        nextClient: undefined,
+    };
+    private lastClient: ILinkedClient = this.rootClient;
+    private nonSummarizerCount = 0;
+    private summarizerCount = 0;
+    private currentClient: ILinkedClient | undefined;
+
+    constructor(quorum: Pick<IQuorum, "getMembers" | "on">) {
+        super();
+        const members = quorum.getMembers();
+        for (const [clientId, client] of members) {
+            this.addClient(clientId, client);
+        }
+
+        quorum.on("addMember", this.addClient);
+        quorum.on("removeMember", this.removeClient);
+
+        this.resetCurrentClient();
+    }
+
+    private readonly addClient = (clientId: string, client: ISequencedClient) => {
+        // Have to undefined-check client.details for backwards compatibility
+        const isSummarizer = client.client.details?.type === summarizerClientType;
+        const newClient: IFullClient = {
+            clientId,
+            sequenceNumber: client.sequenceNumber,
+            isSummarizer,
+            prevClient: undefined,
+            nextClient: undefined,
+        };
+        this.clientMap.set(clientId, newClient);
+
+        if (newClient.isSummarizer) {
+            this.summarizerCount++;
+            this.emit("summarizerChange", this.summarizerCount);
+            return;
+        }
+
+        // Normal case is adding the latest client, which will bypass loop.
+        // Find where it belongs otherwise (this shouldn't happen, assert?).
+        assert(newClient.sequenceNumber > -1, "Negative sequence number not allowed");
+        let currClient = this.lastClient;
+        while (currClient.sequenceNumber > newClient.sequenceNumber) {
+            assert(currClient.prevClient !== undefined, "Previous client should always be defined");
+            // what to do if currClient === this.currentClient
+            currClient = currClient.prevClient;
+        }
+        // Now currClient is the node right before where the new client node should be.
+        newClient.prevClient = currClient;
+        newClient.nextClient = currClient.nextClient;
+        newClient.prevClient.nextClient = newClient;
+        if (newClient.nextClient === undefined) {
+            this.lastClient = newClient;
+        } else {
+            newClient.nextClient.prevClient = newClient;
+        }
+
+        if (this.currentClient === undefined && newClient.nextClient === undefined) {
+            this.currentClient = newClient;
```

Comment:
i don't know what current is yet, but here it looks the be the first last client. which is strange

Reply:
I revamped it a little, so instead of using overly abstract terms like "current client", "last client", and "non-summarizer clients", it is now "elected client", "youngest client", and "eligible clients". So basically this thing tracks ordered clients for purposes of election only. Eligible is the abstraction over interactive, but exists in case we wanted that to mean something else.